harmonica/equivalent_sources/gradient_boosted.py

```diff
@@ -0,0 +1,382 @@
+        data = data.ravel()
+        if weights is not None:
+            weights = weights.ravel()
+        # Build point sources
+        if self.points is None:
+            self.points_ = self._build_points(coordinates)
+        else:
+            self.points_ = vdb.n_1d_arrays(self.points, 3)
+        # Initialize coefficients
+        self.coefs_ = np.zeros_like(self.points_[0])
+        # Fit coefficients through gradient boosting
+        self._gradient_boosting(coordinates, data, weights)
+        return self
+
+    def _gradient_boosting(
+        self, coordinates, data, weights
+    ):  # pylint: disable=too-many-locals
+        """
+        Fit source coefficients through gradient boosting
+        """
+        # Create rolling windows
+        point_windows, data_windows = self._create_windows(coordinates)
+        # Get number of windows
+        n_windows = len(point_windows)
+        # Initialize RMSE array
+        errors = [np.sqrt(np.mean(data ** 2))]
+        # Set weights_chunk to None (will be changed unless weights is None)
+        weights_chunk = None
+        # Initialized the predicted and residue arrays
+        predicted = np.empty_like(data)
+        residue = data.copy()
+        # Iterate over the windows
+        for window in range(n_windows):
+            # Get source and data points indices for current window
+            point_window, data_window = point_windows[window], data_windows[window]
+            # Choose source and data points that fall inside the window
+            points_chunk = tuple(p[point_window] for p in self.points_)
+            coords_chunk = tuple(c[data_window] for c in coordinates)
+            # Choose weights for data points inside the window (if not None)
+            if weights is not None:
+                weights_chunk = weights[data_window]
+            # Compute Jacobian (for sources and data points in current window)
+            jacobian = self.jacobian(coords_chunk, points_chunk)
+            # Fit coefficients of sources with residue points inside window
+            coeffs_chunk = vdb.least_squares(
+                jacobian,
+                residue[data_window],
+                weights_chunk,
+                self.damping,
+            )
+            # Predict field of the sources in the window on every data point
+            predicted[:] = 0
+            predict_numba_parallel(
+                coordinates,
+                points_chunk,
+                coeffs_chunk,
+                predicted,
+                self.greens_function,
+            )
+            # Update the residue
+            residue -= predicted
+            # Add RMS of the residue to the RMSE
+            errors.append(np.sqrt(np.mean(residue ** 2)))
+            # Update source coefficients
+            self.coefs_[point_window] += coeffs_chunk
+        self.errors_ = np.array(errors)
+
+    def _create_windows(self, coordinates, shuffle_windows=True):
```

Comment:
I've added the `shuffle_windows` argument to this function in order to run a test that depends on not shuffling (the one that checks that the correspondence between sources windows and data windows). I'm not planning to make it public since it's a bad idea not to shuffle.

Reply:
Makes perfect sense and it says so in the docstring which is perfect. How about just `shuffle` since the function is already `_create_windows`?

Reply:
Better than that would be import the module instead `from … import`. That’s always preferred to importing single functions since it keeps the namespaces separate.